Local backend and minor fixes

rueckix opened the pull request:
This PR 

- adds a local phonebook (json file) backend that takes precedence over the online directories.
-  fixes incorrect counts in the response message (in case of multiple matching results)
- adds a production WSGI server
- starts using a phonenumber library for parsing and formatting
- returns also the phone number in the _hm_ field

bseibold commented on gigaset/backends/dasoertliche.py:
```diff
@@ -5,33 +5,51 @@
     return []
 
 
+import sys
```
Duplicate

rueckix replied:
done

bseibold commented on gigaset/backends/dasoertliche.py:
```diff
@@ -5,33 +5,51 @@
+    try:
+        with urllib.request.urlopen(url) as resp:
+            return resp.read()
+    except:
```
Please avoid bare `except`, catch at least `Exception`, but better `urllib.error.URLError`. See also https://www.flake8rules.com/rules/E722.html

rueckix replied:
done

bseibold commented on gigaset/backends/dastelefonbuch.py:
```diff
@@ -52,9 +52,13 @@ def _fetch_reverse(phone):
+    try:
+        with urllib.request.urlopen(url) as resp:
+            return resp.read()
+    except:
```
Same as in the other backend, no bare `except` please

rueckix replied:
done

bseibold commented on gigaset/__init__.py:
```diff
@@ -6,9 +6,24 @@
+
 app = Flask(__name__)
+
+os.environ['GIGASET_SETTINGS'] = os.environ.get("GIGASET_SETTINGS", "../settings.py")
```
This should not be necessary if the WSGI environment is configured correctly

rueckix replied:
moved to run.py for testing purposes

bseibold commented on gigaset/__init__.py:
```diff
@@ -6,9 +6,24 @@
 from flask import Flask
 
+import phonenumbers
+from phonenumbers import geocoder
```
This seems to be unused

rueckix replied:
done

bseibold commented on gigaset/__init__.py:
```diff
@@ -6,9 +6,24 @@
+os.environ['GIGASET_SETTINGS'] = os.environ.get("GIGASET_SETTINGS", "../settings.py")
 app.config.from_envvar('GIGASET_SETTINGS')
 
+if app.config['DEBUG']:
```
It would be easier to use `if app.debug:`. In the non-debug case, why not leave the default level?

rueckix replied:
done

bseibold commented on requirements.txt:
```diff
@@ -1 +1,3 @@
-.
```
I don't remember the details how this worked in this old Python version, but the single `.` was intended. It would be great to update the package to something more modern, but in that case the requirements should be moved to `pyproject.toml`. Otherwise just leave this as it is now.

bseibold commented on run.py:
```diff
@@ -1,7 +1,11 @@
 #!/usr/bin/env python3
 
 import gigaset
+import settings
```
Please see https://flask.palletsprojects.com/en/1.1.x/config/
Flask is configured via `app.config.from_envvar()`, the config should then be accessed via `app.config`, and `debug` should already be set. Is is not?

rueckix replied:
using app.config now

bseibold commented on run.py:
```diff
@@ -1,7 +1,11 @@
+
+
 
 if __name__ == '__main__':
```
This whole `main` is just intended for debugging, not for production, therefore having it always in debug mode is fine.

bseibold commented on run.py:
```diff
@@ -2,10 +2,13 @@
-    gigaset.app.run(host='0.0.0.0', port=settings.PORT)
+    
+    
+    serve(gigaset.app, host="0.0.0.0", port=settings.PORT)
```
As mentioned in the other commit, this whole block is intended for debugging only. Please see the documentation of `waitress`, `uwsgi`, or other WSGI servers for details. Personally I was using the following uWSGI config for "production":

```
[uwsgi]
uid = 33
gid = 33
plugin = python3
socket = /var/www/uwsgi/gigaset/gigaset.sock
manage-script-name = true
virtualenv = /var/www/uwsgi/gigaset/venv
chdir = /var/www/uwsgi/gigaset/gigaset
mount = /gigaset=gigaset:app
env = GIGASET_SETTINGS=/var/www/uwsgi/gigaset/settings.py
```

rueckix replied:
reverting the change (except the configurable debug and port configuations.

bseibold commented on settings.example.py:
```diff
@@ -16,10 +16,3 @@
-MQTT_HOST = ""
-MQTT_PORT = 123
-MQTT_USER = ""
-MQTT_PASS = ""
```
Please squash/fixup this into the other commit

rueckix replied:
squashed

bseibold commented:
Hi @rueckix, thanks for your PR, and merry christmas!

I do not have the Gigaset base station in active use any more, so I won't work on this in the future. If you want to take over the maintenance of this project, I could add a link to your repo.

rueckix commented:
> Hi @rueckix, thanks for your PR, and merry christmas!
> 
> I do not have the Gigaset base station in active use any more, so I won't work on this in the future. If you want to take over the maintenance of this project, I could add a link to your repo.

Merci. Happy Holidays! 
Sure, I could take over the maintenance via my fork. In the meantime, thanks for the suggestions and hints! This will further improve the code.

rueckix committed:
+ local phonebook support; +more config options; + changed phone number normalization

updated settings example

enable production WSGI server

updated settings example

code quality

code quality

rueckix committed:
Merge branch 'master' into local-backend

rueckix committed:
merge origin/master
